### PR TITLE
SBOR depth validation

### PR DESCRIPTION
**Pull request by CyonAlexRDX**
Validate SBOR depth in constructors of:
* `Instructions`
* `TransactionIntent`
* `SignedIntent`
* `NotarizedTransaction`

Hide Rust constructors of:
* `CompiledNotarizedIntent`
* `NotarizedTransaction`
* `SignedIntent`
* `TransactionIntent`
* `Instructions`
* `InstructionsSecretMagic`
* `TransactionManifest`

*Thus ensuring we always use Failable constructors ("throwing initializers") of these types - where we have asserted SBOR depth*.

Result:
* All seven instances where I've chosen to `panic` (crash if err) in PR https://github.com/radixdlt/sargon/pull/53 are safe to do so, since we have protected against all failable execution paths.
* `debug_print_compiled_notarized_intent` is no longer throwing, since it is not possible to construct an invalid `CompiledNotarizedIntent`.

Extra:
* UniFFI: exported sample value of TransactionHeader
* Rust: possible to create an address from node_id bytes (random uses it), to be able to create some address on some network with some bytes
* bump `typos` version
* Fix broken Swift post processor (after bump to UniFFI 0.27 the format of inits changed)

### CODE REVIEW

File: src/wrapped_radix_engine_toolkit/low_level/transaction_intent.rs
Comment (CyonAlexRDX) [inline src/wrapped_radix_engine_toolkit/low_level/transaction_intent.rs:53]: cannot use `From<TransactionIntent>` inside the constructor of `TransactionIntent`... duh... so we use this global fn, which is also used by our `impl From<TransactionIntent> for ScryptoIntent` implementation

File: src/wrapped_radix_engine_toolkit/low_level/transaction_intent.rs
Comment (CyonAlexRDX) [inline src/wrapped_radix_engine_toolkit/low_level/transaction_intent.rs:114]: sample values, safe to unwrap.

File: src/wrapped_radix_engine_toolkit/high_level/manifest_building/manifest_assets_transfers.rs
Comment (CyonAlexRDX) [inline src/wrapped_radix_engine_toolkit/high_level/manifest_building/manifest_assets_transfers.rs:65]: this unwraps the failable constructor - which is OK, since this manifest is built by US in Sargon, and we know the SBOR depth.

File: src/wrapped_radix_engine_toolkit/high_level/manifest_building/modify_manifest.rs
Comment (CyonAlexRDX) [inline src/wrapped_radix_engine_toolkit/high_level/manifest_building/modify_manifest.rs:155]: we are not inserting any Instruction with too great an SBOR depth (we insert CallMethod lock fee, and TX Guarantees (WORKTOP ASSERTS))

File: src/wrapped_radix_engine_toolkit/low_level/compiled_notarized_intent.rs
Comment (CyonAlexRDX) [inline src/wrapped_radix_engine_toolkit/low_level/compiled_notarized_intent.rs:16]: hide constructor, ensuring we always go via `TryFrom<ScryptoNotarizedTransaction>`

File: src/wrapped_radix_engine_toolkit/low_level/notarized_transaction.rs
Comment (CyonAlexRDX) [inline src/wrapped_radix_engine_toolkit/low_level/notarized_transaction.rs:5]: Hide constructor so that we ensure we always go via `new` which now is a failable constructor, performing `compile` - checking SBOR depth.

File: src/wrapped_radix_engine_toolkit/low_level/sbor_depth_validation.rs
Comment (CyonAlexRDX) [inline src/wrapped_radix_engine_toolkit/low_level/sbor_depth_validation.rs:14]: Thank you @0xOmarA for providing this!

File: src/wrapped_radix_engine_toolkit/low_level/signed_intent.rs
Comment (CyonAlexRDX) [inline src/wrapped_radix_engine_toolkit/low_level/signed_intent.rs:7]: Hide constructor so that we ensure we always go via `new` which now is a failable constructor, performing `compile_signed_intent_with ` - checking SBOR depth.

File: src/wrapped_radix_engine_toolkit/low_level/transaction_intent.rs
Comment (CyonAlexRDX) [inline src/wrapped_radix_engine_toolkit/low_level/transaction_intent.rs:7]: Hide constructor so that we ensure we always go via `new` which now is a failable constructor, performing `compile_intent_with ` - checking SBOR depth.

File: src/wrapped_radix_engine_toolkit/low_level/transaction_manifest/instructions/instructions.rs
Comment (CyonAlexRDX) [inline src/wrapped_radix_engine_toolkit/low_level/transaction_manifest/instructions/instructions.rs:9]: Hide constructor so that we ensure we always go via `TryFrom<(&[ScryptoInstruction], NetworkID)> ` which calls `scrypto_decompile` this checking SBOR depth

File: apple/Tests/TestCases/RET/TransactionManifestTests.swift
Comment (micbakos-rdx) [inline apple/Tests/TestCases/RET/TransactionManifestTests.swift:44]: Shouldn't those tests be saved in an rtm file too?

**Review by micbakos-rdx — APPROVED** [review]

File: apple/Tests/TestCases/RET/TransactionManifestTests.swift
Comment (Sajjon) [reply apple/Tests/TestCases/RET/TransactionManifestTests.swift:44]: sure, can do it after merge, just wanted to get approval from you and Omar first. Also should generalize file location of RTM files, now there were tied to `execution_summary` (in that folder), but should probably be some layer up, maybe... top level? `$PROJECT_ROOT/examples/transactions` or is it strange to put (new folder) `transactions` next to the iOS example app? speaking of which, is it possible to put the Android example app in that folder? 

if not `$PROJECT_ROOT/examples/transactions`, maybe ``$PROJECT_ROOT/resources/transactions` ? `$PROJECT_ROOT/fixtures/transaction` and also put JSON test vectors in `fixtures/testvectors`??

File: src/wrapped_radix_engine_toolkit/low_level/transaction_manifest/instructions/instructions.rs
Comment (Sajjon) [inline src/wrapped_radix_engine_toolkit/low_level/transaction_manifest/instructions/instructions.rs:51]: @0xOmarA maybe actually it would be enough to create a DUMMY `NotarizedTransaction` here and compile it? If that works, we know that this `Instructions` can be used... so no *need* (but we still can...to be 100% correct and safe) to compile and validate SBOR length in other types (`TransactionIntent`, `SignedIntent`, `NotarizedIntent`, since they all use `Instructions`. 

What do you think?

File: apple/Tests/TestCases/RET/TransactionManifestTests.swift
Comment (micbakos-rdx) [reply apple/Tests/TestCases/RET/TransactionManifestTests.swift:44]: Unfortunatelly the android example app cannot go there. It has to be under the jvm directory with its own build.gradle file.

Regarding the manifests and dat files, the `testvectors` is something I remember @0xOmarA was following with kotlin engine toolkit when we were developing on that repo.

**Review by 0xOmarA — COMMENTED** [review]
Looks good to me! I've left some mostly minor comments and another on the compiled types.

File: src/wrapped_radix_engine_toolkit/low_level/transaction_manifest/instructions/instructions.rs
Comment (0xOmarA) [reply src/wrapped_radix_engine_toolkit/low_level/transaction_manifest/instructions/instructions.rs:51]: My personal preference for this would be to keep the check in this way and to have `Instructions` fairly isolated from the other types and not make assumptions about where it will be used. Other types would have their own depth check that would eventually involve checking the depth of the instructions.

File: src/wrapped_radix_engine_toolkit/low_level/compiled_notarized_intent.rs
Comment (CyonAlexRDX) [reply src/wrapped_radix_engine_toolkit/low_level/compiled_notarized_intent.rs:30]: Right, so the thinking here is that the constructor of `CompiledNotarizedIntent` is hidden in Rust, Swift and Kotlin, so it is not possible to create an instance of it, thus, all instances are valid, since they have been created by calling `compile()` on `NotarizedTransaction` - where `NotarizedTransaction` itself is guaranteed to be valid.

**Comment by 0xOmarA** [comment]
Another comment on this, it seems like there is a common pattern for multiple types where we care about the SBOR depth and want to always ensure that they're below a certain depth. I think that it might be beneficial to do this as a wrapper type where we know that encoding this type is 100% safe and impossible to unwrap. I have an example of such a type here:

```rust
use scrypto::prelude::*;
use std::{borrow::Borrow, ops::Deref};

/// Represents an SBOR type whose depth has been validated. None of the interfaces allow for
/// mutating the inner value as any mutation to it will require another depth check.
#[derive(ManifestSbor, Clone, Debug)]
pub struct DepthValidated<T, const MAX_DEPTH: usize>(T);

impl<T, const MAX_DEPTH: usize> DepthValidated<T, MAX_DEPTH>
where
    T: ManifestEncode + ManifestDecode,
{
    fn new(value: T) -> Option<DepthValidated<T, MAX_DEPTH>> {
        // Attempt to encode with the specified depth.
        drop(manifest_encode_with_depth_limit(&value, MAX_DEPTH).ok()?);

        // Encoding succeeded, value is below the depth limit.
        Some(Self(value))
    }

    fn encode(&self) -> Vec<u8> {
        manifest_encode_with_depth_limit(&self.0, MAX_DEPTH).expect("Check was done in constructor")
    }

    fn decode(slice: &[u8]) -> Option<DepthValidated<T, MAX_DEPTH>> {
        manifest_decode_with_depth_limit(slice, MAX_DEPTH)
            .ok()
            .map(Self)
    }

    fn into_inner(self) -> T {
        self.0
    }
}

impl<T, const MAX_DEPTH: usize> AsRef<T> for DepthValidated<T, MAX_DEPTH> {
    fn as_ref(&self) -> &T {
        &self.0
    }
}

impl<T, const MAX_DEPTH: usize> Borrow<T> for DepthValidated<T, MAX_DEPTH> {
    fn borrow(&self) -> &T {
        &self.0
    }
}

impl<T, const MAX_DEPTH: usize> Deref for DepthValidated<T, MAX_DEPTH> {
    type Target = T;

    fn deref(&self) -> &Self::Target {
        &self.0
    }
}

// Just an example.
type DepthValidatedInstructions = DepthValidated<Vec<InstructionV1>, MANIFEST_SBOR_V1_MAX_DEPTH>;
```

This type gives us some nice properties:

1. It's a standard way of marking that a type has had its SBOR depth checked. 
2. It supports any SBOR depth you may want and you're no longer limited to just the max when doing the encoding.
3. Encoding this type is guaranteed to be 100% safe as it has a small well-defined interface and its well understood that it's immutable.

I mention mutability quickly in one of the comments, but something like this can't be mutable since each mutation would require another encode check.

**Pull request merged**
